File: projects/15-GaleShapley/implementation/GaleShapley.java
import java.util.Arrays;
import java.util.Scanner;
import java.io.File;
import java.io.FileNotFoundException;

public class GaleShapley {

    /**
     * Implements the Gale–Shapley algorithm for the Stable Marriage Problem.
     * 
     * @param n             the number of men (same as the number of women)
     * @param proposerPrefs a 2D array where proposerPrefs[i] is an array of woman
     *                      indices, sorted from most to least preferred by man i.
     * @param acceptorPrefs a 2D array where acceptorPrefs[i] is an array of man
     *                      indices, sorted from most to least preferred by woman i.
     * @return an array "res" where res[i] is the index of the woman matched with
     *         man i.
     */
    public static int[] galeShapley(int n, int[][] proposerPrefs, int[][] acceptorPrefs) {
        // all of the men and women start off as single (free)
        boolean[] menFree = new boolean[n];
        boolean[] womenFree = new boolean[n];
        Arrays.fill(menFree, true);
        Arrays.fill(womenFree, true);

        // res[i] will eventually hold the index of the woman paired with man i.
        int[] res = new int[n];
        // resWomen[i] will hold the index of the man paired with woman i.
        int[] resWomen = new int[n];
        Arrays.fill(res, -1);
        Arrays.fill(resWomen, -1);

        // building womenRef: for each woman and every man that she ranks, store her
        // ranking.
        // lower numbers indicate higher preference.
        int[][] womenRef = new int[n][n];
        for (int w = 0; w < n; w++) {
            int[] prefs = acceptorPrefs[w];
            for (int rank = 0; rank < prefs.length; rank++) {
                int man = prefs[rank];
                womenRef[w][man] = rank;
            }
        }

        // while there is a free man, engage him
        int freeMan = findFreeMan(menFree);
        while (freeMan != -1) {
            // get the free man's preference list.
            int[] manPref = proposerPrefs[freeMan];
            for (int woman : manPref) {
                if (womenFree[woman]) {
                    // if the woman is free, engage her with the man.
                    menFree[freeMan] = false;
                    womenFree[woman] = false;
                    res[freeMan] = woman;
                    resWomen[woman] = freeMan;
                    break;
                } else {
                    // woman is already engaged, so see if she likes the new man potentially
                    int currentMan = resWomen[woman];
                    if (womenRef[woman][freeMan] < womenRef[woman][currentMan]) {
                        // if she prefers the new man, break off her current relationship
                        menFree[currentMan] = true;
                        // engage her with the new man.
                        res[freeMan] = woman;
                        resWomen[woman] = freeMan;
                        menFree[freeMan] = false;
                        break;
                    }
                    // otherwise, she rejects the proposal and the free man moves on to the next
                    // woman.
                }
            }
            // try to find another free man to make a proposal.
            freeMan = findFreeMan(menFree);
        }
        return res;
    }

    /**
     * Helper method to find the index of a free man.
     * 
     * @param menFree an array that indicates whether each man is free.
     * @return the index of a free man, or -1 if none are free.
     */
    private static int findFreeMan(boolean[] menFree) {
        for (int i = 0; i < menFree.length; i++) {
            if (menFree[i]) {
                return i;
            }
        }
        return -1;
    }

    public static void main(String[] args) throws FileNotFoundException {
        String[] testFiles = {
                "io/sample.in.1",
                "io/sample.in.2",
                "io/sample.in.3"
        };
        String[] outputFiles = {
                "io/sample.out.1",
                "io/sample.out.2",
                "io/sample.out.3"
        };
        for (int idx = 0; idx < testFiles.length; idx++) {
            Scanner reader = new Scanner(new File(testFiles[idx]));
            int n = reader.nextInt();
            int[][] proposerPrefs = new int[n][n];
            int[][] acceptorPrefs = new int[n][n];
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    proposerPrefs[i][j] = reader.nextInt();
                }
            }
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    acceptorPrefs[i][j] = reader.nextInt();
                }
            }
            reader.close();
            Scanner outputReader = new Scanner(new File(outputFiles[idx]));
            int[] expectedOutput = new int[n];
            for (int i = 0; i < n; i++) {
                expectedOutput[i] = outputReader.nextInt();
            }
            outputReader.close();

            // Run Gale-Shapley algorithm and assert results
            int[] actualOutput = GaleShapley.galeShapley(n, proposerPrefs, acceptorPrefs);
            assert expectedOutput == actualOutput;
        }
    }
}

File: projects/15-GaleShapley/implementation/gale_shapley.py
from typing import List


# returns an array of n: each index i has a value res[i] which means that
# man i is paired with woman res[i]
def gale_shapley(
    n: int, proposer_prefs: List[List[int]], acceptor_prefs: List[List[int]]
) -> List[int]:
    # boolean arrays to keep track of free proposers and acceptors
    proposer_free = [True] * n
    acceptor_free = [True] * n
    res = [-1] * n
    res_acceptors = [-1] * n

    # dictionary to keep track of the pointers that each proposer keeps to track which acceptor he has proposed to
    proposer_current = {i: 0 for i in range(len(proposer_free))}

    # set to keep track of available proposers with constant time
    available_proposers = set({i for i in range(len(proposer_free))})

    # 2d matrix to store the preferences of the acceptors for fast access for each proposer
    # the key is a tuple (i_a, i_p) where i_a is the index of the acceptor and i_p is the index of the proposer
    # lower the rank, the more preferred the proposer is
    acceptor_reference = {}
    for i_a, pref in enumerate(acceptor_prefs):
        for rank, i_p in enumerate(pref):
            acceptor_reference[(i_a, i_p)] = rank

    while True:
        if available_proposers:
            i_p = available_proposers.pop()
            current_prefs = proposer_prefs[i_p]
            for index in range(proposer_current[i_p], len(current_prefs)):
                i_a = current_prefs[index]
                proposer_current[i_p] = proposer_current[i_p] + 1
                # first case: if this acceptor is free, engage the proposer with the acceptor
                if acceptor_free[i_a]:
                    acceptor_free[i_a] = False
                    proposer_free[i_p] = False
                    res[i_p] = i_a
                    res_acceptors[i_a] = i_p
                    break
                # second case: if the acceptor is not free, check if the new proposer is preferred
                else:
                    old_proposer = res_acceptors[i_a]
                    old_proposer_index = acceptor_reference[(i_a, old_proposer)]
                    new_proposer_index = acceptor_reference[(i_a, i_p)]
                    # if the new proposer is preferred over the old one, redo the
                    # pairing so that the old proposer needs to find a new match
                    # Add the old proposer back to the available proposers set if the old 
                    # proposer needs to find a new match. Otherwise, add the new proposer back 
                    # to the available proposers set if the new proposer is not matched.
                    if new_proposer_index < old_proposer_index:
                        proposer_free[old_proposer] = True
                        proposer_free[i_p] = False
                        res[i_p] = i_a
                        res_acceptors[i_a] = i_p
                        available_proposers.add(old_proposer)
                        break
                    else:
                        available_proposers.add(i_p)

        # if there is no free proposer, return the result
        # it is proven that the algorithm will terminate
        if True not in proposer_free:
            return res


def main():
    for idx, file in enumerate(
        [
            "io/sample.in.1",
            "io/sample.in.2",
            "io/sample.in.3",
        ]
    ):
        with open(file, "r") as f:
            n = int(f.readline().strip())
            proposer_prefs = []
            acceptor_prefs = []
            for i in range(n):
                line = list(map(int, f.readline().strip().split()))
                # print(line)
                proposer_prefs.append(line)
            for i in range(n):
                line = list(map(int, f.readline().strip().split()))
                acceptor_prefs.append(line)
            with open("io/sample.out." + str(idx + 1), "r") as f:
                res = list(map(int, f.readline().strip().split()))
            # print(res, gale_shapley(n, proposer_prefs, acceptor_prefs))
            assert res == gale_shapley(n, proposer_prefs, acceptor_prefs)


if __name__ == "__main__":
    main()
